src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs

```diff
@@ -0,0 +1,219 @@
+            {
+                return new BraceCompletionContext(document, leftToken.GetLocation().SourceSpan.Start, rightToken.GetLocation().SourceSpan.End, caretLocation);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Checks if the already inserted token is a valid opening token at the position in the document.
+        /// By default checks that the opening token is a valid token at the position and not in skipped token trivia.
+        /// </summary>
+        protected virtual Task<bool> IsValidOpenBraceTokenAtPositionAsync(SyntaxToken token, int position, Document document, CancellationToken cancellationToken)
+        {
+            if (token.SpanStart != position)
+            {
+                return SpecializedTasks.False;
+            }
+
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            return Task.FromResult(IsValidOpeningBraceToken(token) && !ParentIsSkippedTokensTrivia(syntaxFactsService, token));
+        }
+
+        /// <summary>
+        /// Returns true when the current position is inside user code (e.g. not strings) and the closing token
+        /// matches the expected closing token for this brace completion service.
+        /// Helper method used by <see cref="AllowOverTypeAsync(BraceCompletionContext, CancellationToken)"/> implementations.
+        /// </summary>
+        protected async Task<bool> AllowOverTypeInUserCodeWithValidClosingTokenAsync(BraceCompletionContext context, CancellationToken cancellationToken)
+        {
+            var tree = await context.Document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFactsService = context.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            return !syntaxFactsService.IsInNonUserCode(tree, context.CaretLocation, cancellationToken)
+                && await CheckClosingTokenKindAsync(context.Document, context.ClosingPoint, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Returns true when the closing token matches the expected closing token for this brace completion service.
+        /// Used by <see cref="AllowOverTypeAsync(BraceCompletionContext, CancellationToken)"/> implementations
+        /// when the over type could be triggered from outside of user code (e.g. overtyping end quotes in a string).
+        /// </summary>
+        protected Task<bool> AllowOverTypeWithValidClosingTokenAsync(BraceCompletionContext context, CancellationToken cancellationToken)
+        {
+            return CheckClosingTokenKindAsync(context.Document, context.ClosingPoint, cancellationToken);
+        }
+
+        protected static bool ParentIsSkippedTokensTrivia(ISyntaxFactsService syntaxFactsService, SyntaxToken token)
+            => token.Parent == null || syntaxFactsService.IsSkippedTokensTrivia(token.Parent);
```

Comment:
method body doesn't match method name.  not sure how relevant that is.  is this copied from previous code?